### PR TITLE
Json ld in html

### PR DESCRIPTION
**Pull request by gkellogg**
* Spec text for running API methods with HTML input. Adds `extractAllScripts` option and adds `contentType` accessor to `RemoteDocument`.
* HTML tests for expand, compact, flatten and toRdf.

For w3c/json-ld-syntax#57.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/json-ld-api/pull/50.html" title="Last updated on Nov 13, 2018, 10:17 PM GMT (6a6171a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/json-ld-api/50/804113b...6a6171a.html" title="Last updated on Nov 13, 2018, 10:17 PM GMT (6a6171a)">Diff</a>

### CODE REVIEW

**Comment by gkellogg** [comment]
> Does this require conformant JSON-LD 1.1 processors to handle the embedded-in-HTML scenario?

It seems to me that we discussed creating optional subsets of the spec, and other than for Framing, rejected the idea as being poor for interoperability. Of course, we can discuss it again. As I recall, the barrier for implementations to include HTML parsing seemed low enough that all implementations should be expected to do this. (@azaroth42 may remember more, I think he's the one that made that point).

If we do want to go there, the tests should be removed from the expand/compact/flatten/toRdf manifests and a new html-manifest should be created, so that it doesn't look like scattered failures.

> one that comes with protocol "kit" to dereference comment URL's for contexts (http, https, etc)--i.e. includes "requests" plumbing, but no HTML parsing

I recall that in discussing URL options for documentation, the ability to use JSON-LD embedded in HTML seemed like the best way to do such documentation.

File: tests/expand/h014-in.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<html>
+  <head>
+    <script type="application/ld+json">
+    <!--
+    {
+      "@context": {
+        "foo": {"@id": "http://example.com/foo", "@container": "@list"}
+      },
+      "foo": [{"@value": "<!-- -->"}]
```
[the remark this answers is not in the document]
Comment (gkellogg) [reply tests/expand/h014-in.html]: Certainly, HTML embedded in HTML contained in an HTML script would be one such use. Also, various HTML validators complain when script elements contain things that it doesn't expect, and placing comments around the entire JSON-LD block is one solution for this.

Intermingling without escapes is not supported, to be intermingled, they must be escaped. This could be using `<\!-- --\>` or `&lt;!-- --&gt;`. Easiest thing is to comment out the whole block, and deal specifically with any embedded XML comments rather than entity encode the contents; either would work, though.

File: tests/expand/h015-in.html [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<html>
+  <head>
+    <script type="application/ld+json">
+    <!--
```
[the remark this answers is not in the document]
Comment (gkellogg) [reply tests/expand/h015-in.html]: That would be invalid per our text, as after unescaping, "this is for SEO" would not be valid JSON.

[the remark this answers is not in the document]
Comment (BigBlueHat) [reply tests/expand/h015-in.html]: Right. Nothing we can change...just didn't know if we should have a test that looked like that-kind-of-broken. 😄

**Comment by BigBlueHat** [comment]
@gkellogg as we consider other deployments like Web of Things (i.e. JSON-LD in a light bulb), we might want to consider revisiting these growing processing requirements being made on *all* processors.

Related to this, we should weigh whether `Content-Type: text/html` responses are to be treated immediately as a valid JSON-LD "encoding" or if this is a special scenario from which JSON-LD may be extracted (using defined steps per this PR) and *then* fed to a "JSON-LD processor."

Seems modulerizing it a bit (see [my list above](https://github.com/w3c/json-ld-api/pull/50#pullrequestreview-174070235)) could actually widen the use of JSON-LD and narrow the scope of "interop." Maybe. 😃

File: tests/expand/h015-in.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (gkellogg) [reply tests/expand/h015-in.html]: expand/h017-in.html looks for invalid JSON-LD, but not inside a comment. Probably easiest to modify this to the following:

```html
<html>
  <head>
    <script type="application/ld+json">
    <!-- foo -->
    </script>
  </head>
</html>
```

**Comment by gkellogg** [comment]
> might want to consider revisiting these growing processing requirements being made on _all_ processors.

Sounds like a worthy topic of a teleconference, but I think we did already consider it.

**Comment by gkellogg** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2018/2018-11-16-json-ld#section3).

- `Resolution #2: merge open HTML related PRs #93 and #68 and #50 after adding “At Risk” (or similar terminology) to present that things are not finalized `
<details><summary><i class="differentiate">View the transcript</i></summary>
What is ‘base’ for embedded json-ld?<br/>
<blockquote>
  <p><em>Ivan Herman:</em> link: <a href="https://github.com/w3c/json-ld-syntax/issues/23">https://github.com/w3c/json-ld-syntax/issues/23</a></p>
</blockquote>

<blockquote>
  <p><em>Ivan Herman:</em> Link of the PR: <a href="https://github.com/w3c/json-ld-syntax/pull/93">https://github.com/w3c/json-ld-syntax/pull/93</a></p>
</blockquote>

<blockquote>
  <p><em>Benjamin Young:</em> <a href="https://github.com/w3c/json-ld-syntax/pull/68">https://github.com/w3c/json-ld-syntax/pull/68</a></p>
</blockquote>

<blockquote>
  <p><em>Gregg Kellogg:</em> <a href="https://github.com/w3c/json-ld-api/pull/50">https://github.com/w3c/json-ld-api/pull/50</a></p>
</blockquote>

<p><strong>Benjamin Young:</strong> we discussed that one at tpac<br />
… we sent it in for TAG review, and they basically widened the scope</p>

<p><strong>Gregg Kellogg:</strong> there are 2 open PRs<br />
… 1) basic support for json-ld in html<br />
… 2) PR-93 adds text to specifically add text to add html as base<br />
… in the API spec, it’s PR-50</p>

<p><strong>Adam Soroka:</strong> quick question, what are we expected to do with their comments?<br />
… shall we respond?</p>

<p><strong>Ivan Herman:</strong> what they propose is interesting but beyond our charter<br />
… this would elevate json-ld<br />
… but yeah beyond our charter<br />
… I would say this is something the CG has to pick up<br />
… and we can cross the bridge at some time, but if this is realistic from a manpower perspective I dont know</p>

<blockquote>
  <p><em>Benjamin Young:</em> <a href="https://github.com/w3c/json-ld-syntax/pull/68#issuecomment-438015329">https://github.com/w3c/json-ld-syntax/pull/68#issuecomment-438015329</a></p>
</blockquote>

<p><strong>Ivan Herman:</strong> regarding the PR-93, there is some stuff about having XML</p>

<p><strong>Benjamin Young:</strong> the thing I just linked shows how script tags affect html parsing<br />
… it’s syntactically correct json-ld</p>

<p><strong>Gregg Kellogg:</strong> what I did in the PR-68 I call out specifics on how to handle those blocks if the media type is application/json<br />
… I think I’ve taken in the specifics on how content of script tags has to be handled and adjusted in for our needs<br />
… we asked specific questions to TAG, got an answer but they kinda got a bit over enthusiastic<br />
… out of this needs to come something that improves web platform</p>

<p><strong>Benjamin Young:</strong> the HTML comments stuff as really bothered me since I’ve read it<br />
… but it seems to primarily affect only HTML parsing<br />
… question is how much of this we need to have in the spec<br />
… json-ld in script tags vs “raw” json-ld<br />
… both have totally different escaping rules and what not.. and none of that has something to do with html base</p>

<p><strong>Ivan Herman:</strong> for the comment storing, the whole section is a normative thing<br />
… I have the impression this is an HTML problem<br />
… which we should certainly mention, but maybe not as part of a normative section</p>

<blockquote>
  <p><em>Benjamin Young:</em> HTML5 spec level text about parsing <code class="highlighter-rouge">&lt;script&gt;</code> tags <a href="https://www.w3.org/TR/html5/semantics-scripting.html#restrictions-for-contents-of-script-elements">https://www.w3.org/TR/html5/semantics-scripting.html#restrictions-for-contents-of-script-elements</a></p>
</blockquote>

<p><strong>Ivan Herman:</strong> we should officially answer to the TAG and will officially add to the standard what they said about base<br />
… and that we try to get the CG involved</p>

<blockquote>
  <p><em>Adam Soroka:</em> +1</p>
</blockquote>

<p><strong>Gregg Kellogg:</strong> comments in html and escaping.. it depends on the encoding<br />
… we don’t need to give guidance on how to handle this</p>

<blockquote>
  <p><em>Ivan Herman:</em> <a href="https://pr-preview.s3.amazonaws.com/w3c/json-ld-syntax/pull/68.html#ex-103-embedding-json-ld-in-html-with-comments">https://pr-preview.s3.amazonaws.com/w3c/json-ld-syntax/pull/68.html#ex-103-embedding-json-ld-in-html-with-comments</a></p>
</blockquote>

<p><strong>Ivan Herman:</strong> it has to be valid json-ld<br />
… that’s invalid json</p>

<p><strong>Gregg Kellogg:</strong> that’s something you see quite often</p>

<blockquote>
  <p><em>Benjamin Young:</em> ivan: because <a href="https://www.w3.org/TR/html5/semantics-scripting.html#restrictions-for-contents-of-script-elements">https://www.w3.org/TR/html5/semantics-scripting.html#restrictions-for-contents-of-script-elements</a></p>
</blockquote>

<p><strong>Gregg Kellogg:</strong> comments are often used just to make sure there are no other issues embedded in the script elements that would cause any issues</p>

<p><strong>Benjamin Young:</strong> I did quite some digging on that issue<br />
… the DOM parsing is only concerned with what’s inside the tags<br />
… it’s just treated as raw string<br />
… if there’s something inside the json-ld an html parser would choke on<br />
… the json-ld would need to be treated in such a way such that an html parser wouldn’t choke on it</p>

<p><strong>Pierre-Antoine Champin:</strong> one crazy idea by looking at the json-ld embedded in html comments: you could add a js comment in front of the html comment, making it valid javascript<br />
… it could become technically correct</p>

<p><strong>Benjamin Young:</strong> sadly it wouldn’t<br />
… it would continue being parsed<br />
… we could make it our own parsing space<br />
… the question is how far the parser gets before it finds the ending script tag</p>

<p><strong>Gregg Kellogg:</strong> the json-ld would not be allowed to contain anything that could be interpreted as html and/or html comments<br />
… not really feasible and also not helping our mission<br />
… I’ve outlined multiple approaches to tackle this<br />
… there are only a few cases where json-ld would contain things that resemble comments</p>

<p><strong>Harold Solbrig:</strong> why is this an json-ld issue but not a javascript issue?</p>

<p><strong>Gregg Kellogg:</strong> [explains why it isn’t]</p>

<p><strong>Gregg Kellogg:</strong> it did some test cases for this, exploring corner cases we know of<br />
… I don’t know how to move forward unless addressing at least some of the stuff from the PR</p>

<p><strong>Gregg Kellogg:</strong> it describes script tags and data blocks are a subset</p>

<p><strong>Benjamin Young:</strong> what’s breaking it, is the potential of one to too early close the script tag<br />
… so this would need somehow being taken care of<br />
… the risk is, the json-ld could contain content that jacks up the html it’s contained in</p>

<p><strong>Ivan Herman:</strong> is it so horrible to say, if I put json-ld in a script tag I’m supposed to escape anything that html would need to have escaped<br />
… thus a json-ld parser would have to do the unescaping<br />
… but you are in HTML regardless.. so</p>

<p><strong>Gregg Kellogg:</strong> for someone who’s actually looking at the source, those entities become rather annoying</p>

<p><strong>Ivan Herman:</strong> realistically, I don’t know how often this would happen</p>

<p><strong>Benjamin Young:</strong> the escaping issue is very similar of putting json-ld inside a text env.</p>

<p><strong>Ivan Herman:</strong> I think it’s perfectly reasonable to accept both PRs, close the issue<br />
… and open a new issue on the specific problem</p>

<p><strong>Gregg Kellogg:</strong> it’s a editor’s draft not a working draft</p>

<p><strong>Ivan Herman:</strong> we would open a issue right away</p>

<p><strong>Benjamin Young:</strong> I would only +1 this, if we add a big red AT RISK disclaimer</p>

<p><strong>Ivan Herman:</strong> a lot of very important things are pending for now<br />
… I think it’s an edge case</p>

<p><strong>Adam Soroka:</strong> I don’t think we should use a phrase like “AT RISK” but more something along the lines of “will be part of the final spec but might undergo some changes”</p>

<p><strong>Ivan Herman:</strong> we cannot commit ourselves to having always consistent editor’s drafts</p>

<p><strong>Benjamin Young:</strong> I’m not sure we have reached consensus on all the things contained</p>

<p><strong>Gregg Kellogg:</strong> I cannot work on other open issues</p>

<p><strong>Pierre-Antoine Champin:</strong> what about a parameter on the media type hinting at having to do unescaping? (like <code class="highlighter-rouge">application/ld+json;escaped=html</code>)</p>

<blockquote class="proposed_resolution">
  <p><strong>Proposed resolution: merge open HTML related PRs #93 and #68 and #50 after adding “At Risk” (or similar terminology) to present that things are not finalized</strong> <em>(Benjamin Young)</em></p>
</blockquote>

<blockquote>
  <p><em>Benjamin Young:</em> +1</p>
</blockquote>

<blockquote>
  <p><em>Adam Soroka:</em> +1</p>
</blockquote>

<p><strong>Ivan Herman:</strong> what does “that” mean?</p>

<blockquote>
  <p><em>Gregg Kellogg:</em> +1</p>
</blockquote>

<p><strong>Benjamin Young:</strong> I don’t want to have stuff merged without reaching consensus<br />
… but I could live with having it marked as being “at risk” or similar</p>

<p><strong>Ivan Herman:</strong> putting things that are already done “at risk” would be going backwards<br />
… opening a new issue that highlights things that are still being discussed would be ok<br />
… having the feeling that ~90% are done</p>

<p><strong>Adam Soroka:</strong> I have to generally agree with ivan</p>

<blockquote>
  <p><em>Ivan Herman:</em> +1</p>
</blockquote>

<p><strong>Adam Soroka:</strong> it seems for me very unlikely that we would stop talking about it</p>

<blockquote>
  <p><em>David Newbury:</em> +1</p>
</blockquote>

<blockquote>
  <p><em>Harold Solbrig:</em> +1</p>
</blockquote>

<p><strong>Benjamin Young:</strong> I’m fine with merging those</p>

<blockquote>
  <p><em>Simon Steyskal:</em> +1</p>
</blockquote>

<blockquote id="resolution2" class="resolution">
  <p><strong><em>Resolution #2: merge open HTML related PRs #93 and #68 and #50 after adding “At Risk” (or similar terminology) to present that things are not finalized</em></strong></p>
</blockquote>

<blockquote>
  <p><em>Pierre-Antoine Champin:</em> +1</p>
</blockquote>


</details>
